### PR TITLE
feat: double password entry

### PR DESCRIPTION
**Pull request by danielgent**
Adding @frankjonatanschultz for visibility of the technical debt nightmare that Markus and I have created :smile:

### CODE REVIEW

File: src/routes/tsx/routes.tsx [hunk 1]
```diff
@@ -29,23 +29,30 @@ export async function postResetPassword(
   }
 
   let password = "";
+  let reEnterPassword = "";
 
   if (contentType === "application/json") {
     // in our tests we are POSTing up JSON, which previously worked
     const json = await ctx.req.json();
     password = json.password;
+    reEnterPassword = json["re-enter-password"];
```
Comment (danielgent) [inline src/routes/tsx/routes.tsx:38]: this fields actually matches auth0.  I would be happy to completely copy Auth0 on this route :sunglasses:  Naming everything the same

image


Interestingly the password strength validation is done clientside, but the validation of matching passwords is done serverside (in a normal form POST)

I'd be happy to leave everything serverside here for now

File: src/routes/tsx/routes.tsx [hunk 2]
```diff
@@ -54,16 +61,17 @@ export async function postResetPassword(
...
-  }
   if (!state) {
     throw new HTTPException(400, { message: "State required" });
   }
   if (!code) {
     throw new HTTPException(400, { message: "Code required" });
   }
 
+  if (password !== reEnterPassword) {
+    return ctx.html(<ResetPasswordPage error="Passwords do not match" />, 400);
```
Comment (danielgent) [inline src/routes/tsx/routes.tsx:72]: Check it out it's React but on the server :sunglasses:  Only on the server

File: test/integration/flows/password.spec.ts [hunk 3]
```diff
@@ -763,6 +763,7 @@ describe("password-flow", () => {
       const resetPassword = await anyClient.u["reset-password"].$post({
         json: {
           password: "New-password-1234!",
+          "re-enter-password": "New-password-1234!",
```
Comment (danielgent) [inline test/integration/flows/password.spec.ts:766]: these changes were needed of course for the tests to pass!

I could also add tests to read the HTML we get back from the reset password page to test it... (although the dream would be using playwright but that would make it way slower)

File: src/routes/tsx/routes.tsx (hunk 1 above)
Comment (markusahlstrand) [reply src/routes/tsx/routes.tsx:38]: Yes, steal with pride!

File: src/routes/tsx/routes.tsx [hunk 4]
```diff
@@ -29,23 +29,30 @@ export async function postResetPassword(
...
   }
 
   if (contentType === "application/x-www-form-urlencoded") {
     // but in the browser we are doing a POST with form data
     const body = await ctx.req.parseBody();
 
     const bodyPassword = body.password;
-    if (typeof bodyPassword !== "string") {
+    const bodyReEnterPassword = body["re-enter-password"];
+    if (
```
Comment (markusahlstrand) [inline src/routes/tsx/routes.tsx:47]: Zod for the win!

**Review by markusahlstrand — APPROVED** [review]
LGTM!

**Pull request merged**